src/elbo.jl

```diff
@@ -0,0 +1,235 @@
+"Likelihoods which take a scalar (or vector of scalars) as input and return a single scalar."
+ScalarLikelihood = Union{BernoulliLikelihood,PoissonLikelihood,GaussianLikelihood}
+
+
+abstract type ExpectationMethod end
+struct Default <: ExpectationMethod end
+struct Analytic <: ExpectationMethod end
+
+struct Quadrature <: ExpectationMethod
+    n_points::Int
+end
+Quadrature() = Quadrature(20)
+
+struct MonteCarlo <: ExpectationMethod
+    n_samples::Int
+end
+MonteCarlo() = MonteCarlo(20)
+
+"""
+    elbo(fx::FiniteGP, y::AbstractVector{<:Real}, fz::FiniteGP, q::AbstractMvNormal; n_data=length(y), method=Default())
+
+Compute the Evidence Lower BOund from [1] for the process `fx.f` where `y` are
+observations of `fx`, pseudo-inputs are given by `z = fz.x` and `q(u)` is a
+variational distribution over inducing points `u = f(z)`.
+
+`method` selects which method is used to calculate the expected loglikelihood in
+the ELBO. The options are: `Default()`, `Analytic()`, `Quadrature()` and
+`MonteCarlo()`. For likelihoods with an analytic solution, `Default()` uses this
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function elbo(
```

Comment:
To answer your question what is assumed to be computed here if we have no information about the likelihood?

Reply:
If I understand correctly, we implicitly use the Gaussian likelihood, so one could simply call `elbo(LatentFiniteGP(fx, GaussianLikelihood()), y, fz, q; ...)` or something similar

Reply:
Yes, exactly, but what if `fx.Σy` is not homoscedastic - I can't figure out a good way to do something like `elbo(LatentFiniteGP(fx, GaussianLikelihood(diag(fx.Σy)), y, fz, q; ...)` if that makes sense

Reply:
Unless we just assume it's homoscedastic for now and use `GaussianLikelihood(Σy[1])` as I already do for MC/quadrature for the Gaussian likelihood

Reply:
Could definitely do that -- would make sense to me.

Might be an idea to dispatch on `fx.Σy` being a `Diagonal{<:Real, <:Fill}` or something though to prevent silent failure.

Reply:
Using heteroskedastic likelihoods is definitely not standard, I think we can ignore this case as a default.

Reply:
Alright, I'll do that. @willtebbutt should I maybe just throw a warning if the matrix isn't explicitly `Diagonal`? otherwise it seems a bit overzealous to only dispatch on that?

Reply:
Good idea. You could also add a method which throws an informative error if you get something other than a `Diagonal{<:Real, <:Fill}`.

Reply:
> Using heteroskedastic likelihoods is definitely not standard, I think we can ignore this case as a default.

Allowing for heteroskedastic likelihoods is actually very relevant in practice, it would be really good to keep that use-case in mind in the design instead of inadvertently making it really hard. The key difference to "scalar" likelihoods is that now there are several latent outputs for each observation.

So this requires multi-output GPs - but for example with mixing kernels the number of "latent outputs" (what the likelihood sees) can be different from the number of "latent GPs" (what we parametrize with q(u))...

Reply:
I've changed `elbo` to dispatch on the type of `fx.Σy` - I'm not 100% sure about the way I've done it and the style though, so feedback would be v appreciated.